nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs

```diff
@@ -331,7 +334,187 @@ internal static async Task<bool> DependenciesAreCoherentAsync(string repoRoot, s
         }
     }
 
-    internal static async Task<Dependency[]?> ResolveDependencyConflicts(string repoRoot, string projectPath, string targetFramework, Dependency[] packages, Logger logger)
+    internal static bool UseNewDependencySolver()
+    {
+        if (Environment.GetEnvironmentVariable("UseNewNugetPackageResolver") == "true")
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    internal static async Task<Dependency[]?> ResolveDependencyConflicts(string repoRoot, string projectPath, string targetFramework, Dependency[] packages, Dependency[] update, Logger logger)
+    {
+        if (UseNewDependencySolver())
+        {
+            return await ResolveDependencyConflictsNew(repoRoot, projectPath, targetFramework, packages, update, logger);
+        }
+        else
+        {
+            return await ResolveDependencyConflictsOld(repoRoot, projectPath, targetFramework, packages, logger);
+        }
+    }
+
+    internal static async Task<Dependency[]?> ResolveDependencyConflictsNew(string repoRoot, string projectPath, string targetFramework, Dependency[] packages, Dependency[] update, Logger logger)
+    {
+        var tempDirectory = Directory.CreateTempSubdirectory("package-dependency-coherence_");
+        PackageManager packageManager = new PackageManager(repoRoot, projectPath);
+
+        try
+        {
+            string tempProjectPath = await CreateTempProjectAsync(tempDirectory, repoRoot, projectPath, targetFramework, packages);
+            var (exitCode, stdOut, stdErr) = await ProcessEx.RunAsync("dotnet", $"restore \"{tempProjectPath}\"", workingDirectory: tempDirectory.FullName);
+
+            // Add Dependency[] packages to List<PackageToUpdate> existingPackages
+            List<PackageToUpdate> existingPackages = packages
+            .Select(existingPackage => new PackageToUpdate
+            {
+                packageName = existingPackage.Name,
+                currentVersion = existingPackage.Version
+            })
+            .ToList();
+
+            // Add Dependency[] update to List<PackageToUpdate> packagesToUpdate
+            List<PackageToUpdate> packagesToUpdate = update
+            .Where(package => package.Version != null)
+            .Select(package => new PackageToUpdate
+            {
+                packageName = package.Name,
+                newVersion = package.Version.ToString()
+            })
+            .ToList();
+
+            foreach (var existing in existingPackages)
+            {
+                var x = packagesToUpdate.Where(p => p.packageName == existing.packageName);
```

Comment:
nit: could you please use a more descriptive variable other than `x`?

Reply:
Haha, this one is my fault.  We were doing some pair coding and I suggested `x` just so we could get it working, then we'd give it a better name later.  I guess later is now.

Unrelated, but on this same line, NuGet package names are case insensitive; this should be `string.Equals(p.packageName, existing.packageName, StringComparison.OrdinalIgnoreCase)`